log4j-api/src/main/java/org/apache/logging/log4j/spi/UnmodifiableArrayBackedMap.java

```diff
@@ -0,0 +1,486 @@
+            return new UnmodifiableEntry(index++);
+        }
+    }
+
+    /**
+     * Simple Entry set, providing a reference to UnmodifiableEntryIterator and
+     * blocking modifications.
+     */
+    private class UnmodifiableEntrySet extends AbstractSet<Map.Entry<String, String>> {
+
+        @Override
+        public boolean add(Entry<String, String> e) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean addAll(Collection<? extends Entry<String, String>> c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Iterator<Entry<String, String>> iterator() {
+            return new UnmodifiableEntryIterator();
+        }
+
+        @Override
+        public int size() {
+            return numEntries;
+        }
+    }
+
+    private static final long serialVersionUID = 6849423432534211514L;
+
+    public static final UnmodifiableArrayBackedMap EMPTY_MAP = new UnmodifiableArrayBackedMap(0);
+
+    private final Object[] backingArray;
+    private int numEntries;
+
+    private static final int NUM_FIXED_ARRAY_ENTRIES = 1;
+
+    private static int getArrayIndexForKey(int entryIndex) {
+        return 2 * entryIndex + NUM_FIXED_ARRAY_ENTRIES;
+    }
+
+    private static int getArrayIndexForValue(int entryIndex) {
+        return 2 * entryIndex + 1 + NUM_FIXED_ARRAY_ENTRIES;
+    }
+
+    private UnmodifiableArrayBackedMap(int capacity) {
+        this.backingArray = new Object[capacity * 2 + 1];
+        this.backingArray[0] = 0;
+    }
+
+    static UnmodifiableArrayBackedMap getInstance(Object[] backingArray) {
+        if (backingArray == null || backingArray.length == 1) {
+            return EMPTY_MAP;
+        } else {
+            return new UnmodifiableArrayBackedMap(backingArray);
+        }
+    }
+
+    private UnmodifiableArrayBackedMap(Object[] backingArray) {
+        this.numEntries = (backingArray == null ? 0 : (int) backingArray[0]);
+        this.backingArray = backingArray;
+    }
+
+    UnmodifiableArrayBackedMap(UnmodifiableArrayBackedMap other) {
+        this.backingArray = other.backingArray;
+        this.numEntries = other.numEntries;
+    }
+
+    private void add(String key, String value) {
+        backingArray[getArrayIndexForKey(numEntries)] = key;
+        backingArray[getArrayIndexForValue(numEntries)] = value;
+        numEntries++;
+    }
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException("Instance cannot be cleared, reuse EMPTY_MAP instead.");
+    }
+
+    /**
+     * Scans the array to find a matching key. Linear performance.
+     */
+    @Override
+    public boolean containsKey(Object key) {
+        int hashCode = key.hashCode();
+        for (int i = 0; i < numEntries; i++) {
+            if (backingArray[getArrayIndexForKey(i)].hashCode() == hashCode
+                    && backingArray[getArrayIndexForKey(i)].equals(key)) {
```

Comment:
My JMH test shows that hashcode is faster, but that's because the particular strings are reused, and each String instance caches its hashcode.  In a situation with un-cached hashcodes, you're right that equals() is the better choice.

Our applications use fixed/reusable/cached strings but they might not be the general case.  Which direction should we bias?

equals() is still fine on our apps, no worries if that's your preference.

Reply:
Good point, I think most strings will be compile-time constants, so let's keep `hashCode`.